### PR TITLE
MRG: Add "description" entity to BIDSPath

### LINKED ISSUE #449: Adding desc keyword to BIDSPath and get_entities_from_fname?
## Describe the problem
Just starting a convo: it seems that the initial Derivatives PR has been merged into the BIDS specification (https://bids-specification.readthedocs.io/en/stable/05-derivatives/01-introduction.html). 

Most explicitly, they seem to add this `desc-[label]` keyword argument to differentiate different files.

#### Describe your solution
Add `desc` keyword to `make_bids_basename`.

#### Describe possible alternatives
A separate function called `make_derivatives_basename`?

#### Additional context
I'm wondering if there are other ways we can bottleneck how people setup their derivatives folders that adheres to the initial specification? E.g. folder naming, etc.

### LINKED ISSUE #1049: Add "description" entity to BIDSPath
See https://openneuro.org/datasets/ds003775/versions/1.0.0/file-display/derivatives:cleaned_data:sub-001:ses-t1:eeg:sub-001_ses-t1_task-resteyesc_desc-cleaned_eeg.set

File is called `sub-001_ses-t1_task-resteyesc_desc-cleaned_eeg.set`
What is this `desc` thing? Currently I cannot load this file with mne-bids as we have no such thing...

ping @sappelhoff @hoechenberger

### PR DESCRIPTION
**Pull request by hoechenberger**
Fixes #1049
Fixes #449

It seems to be working already:

```python
# %%
import mne
import mne_bids

sample_dir = mne.datasets.sample.data_path()
raw_path = sample_dir / 'MEG' / 'sample' / 'sample_audvis_raw.fif'
raw = mne.io.read_raw_fif(raw_path, preload=False)

bp = mne_bids.BIDSPath(
    subject='foo',
    task='bar',
    description='baz',
    suffix='meg',
    extension='.fif',
    datatype='meg',
    root='/tmp',
)
bp_written = mne_bids.write_raw_bids(
    raw=raw,
    bids_path=bp,
    overwrite=True
)
mne_bids.read_raw_bids(bp_written)
```

@sappelhoff It's unclear to me where exactly to inject the `desc` entity in the BIDS path. For now, I'm putting it at the very end, before `split` and `suffix`. I couldn't find anything in the BIDS spec to help here.

Screen Shot 2022-08-27 at 10 25 14

@agramfort Please feel free to test and also to finalize ;)

Merge checklist
---------------

Maintainer, please confirm the following before merging.
If applicable:

- [ ] All comments are resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/main/doc/whats_new.rst) is updated
- [ ] New contributors have been added to [CITATION.cff](https://github.com/mne-tools/mne-bids/blob/main/CITATION.cff)
- [x] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

File: mne_bids/path.py [hunk 1]
```diff
@@ -179,6 +178,10 @@ class BIDSPath(object):
         Corresponds to "space".
         Note that valid values for ``space`` must come from a list
         of BIDS keywords as described in the BIDS specification.
+    description : str | None
+        This corresponds to the BIDS entity  ``desc``. It is used to provide
```
Comment (hoechenberger) [inline mne_bids/path.py:182]:
```suggestion
        This corresponds to the BIDS entity ``desc``. It is used to provide
```

File: mne_bids/path.py [hunk 2]
```diff
@@ -288,36 +291,38 @@ class BIDSPath(object):
...
-            ('session', self.session),
-            ('task', self.task),
-            ('acquisition', self.acquisition),
-            ('run', self.run),
-            ('processing', self.processing),
-            ('space', self.space),
-            ('recording', self.recording),
-            ('split', self.split)
-        ])
+        return {
```
Comment (hoechenberger) [inline mne_bids/path.py:314]: dict insertion order is preserved in Python 3.7+, so no need to have an OrderedDict anymore

**Comment by hoechenberger** [comment]
Needs to be added to get_entities_from_fname() (???) too

File: mne_bids/path.py [hunk 3]
```diff
@@ -179,6 +178,12 @@ class BIDSPath(object):
         Corresponds to "space".
         Note that valid values for ``space`` must come from a list
         of BIDS keywords as described in the BIDS specification.
+    description : str | None
+        This corresponds to the BIDS entity ``desc``. It is used to provide
+        additional information for  derivative data, e.g., preprocessed data
```
Comment (agramfort) [inline mne_bids/path.py:183]:
```suggestion
        additional information for derivative data, e.g., preprocessed data
```

**Comment by agramfort** [comment]
I would say yes to be consistent

**Comment by hoechenberger** [comment]
Yep absolutely, I just wasn't sure about the exact name of that function :)

**Comment by hoechenberger** [comment]
Doc build is broken due to recent changes upstream in MNE.

**Comment by agramfort** [comment]
@alexrockhill can you have a look the doc build issue? It’s related to seeg example 🙏

**Comment by sappelhoff** [comment]
> @sappelhoff It's unclear to me where exactly to inject the desc entity in the BIDS path. For now, I'm putting it at the very end, before split and suffix. I couldn't find anything in the BIDS spec to help here.

see: https://github.com/bids-standard/bids-specification/blob/b3afa2ada5faaebeef9c48e6074a6fa5174289a3/src/schema/rules/entities.yaml#L1-L31

**Comment by hoechenberger** [comment]
Wow, how did you even find that 😅 but perfect

**Comment by hoechenberger** [comment]
Actually no it's not good. Why is split not last? Having it not be last doesn't make any sense to me.

File: mne_bids/path.py [hunk 4]
```diff
@@ -1849,14 +1875,15 @@ def _filter_fnames(fnames, *, subject=None, session=None, task=None,
     rec_str = f'_rec-{recording}' if recording else r'(|_rec-([^_]+))'
     space_str = f'_space-{space}' if space else r'(|_space-([^_]+))'
     split_str = f'_split-{split}' if split else r'(|_split-([^_]+))'
+    desc_str = f'_desc-{space}' if description else r'(|_desc-([^_]+))'
```
Comment (hoechenberger) [inline mne_bids/path.py:1878]:
```suggestion
    desc_str = f'_desc-{desc}' if description else r'(|_desc-([^_]+))'
```

File: mne_bids/path.py [hunk 5]
```diff
@@ -1849,14 +1875,15 @@ def _filter_fnames(fnames, *, subject=None, session=None, task=None,
...
+    desc_str = f'_desc-{desc}' if description else r'(|_desc-([^_]+))'
     suffix_str = (f'_{suffix}' if suffix
                   else r'_(' + '|'.join(ALLOWED_FILENAME_SUFFIX) + ')')
     ext_str = extension if extension else r'.([^_]+)'
 
     regexp = (
         leading_path_str +
         sub_str + ses_str + task_str + acq_str + run_str + proc_str +
-        rec_str + space_str + split_str + suffix_str + ext_str
+        rec_str + space_str + desc_str + split_str + suffix_str + ext_str
```
Comment (hoechenberger) [inline mne_bids/path.py:1886]:
```suggestion
        rec_str + space_str + split_str + desc_str + suffix_str + ext_str
```

File: mne_bids/path.py [hunk 6]
```diff
@@ -1654,6 +1675,7 @@ def get_entity_vals(root, entity_key, *, ignore_subjects='emptyroom',
     ignore_spaces = _ensure_tuple(ignore_spaces)
     ignore_acquisitions = _ensure_tuple(ignore_acquisitions)
     ignore_splits = _ensure_tuple(ignore_splits)
+    ignore_splits = _ensure_tuple(ignore_splits)
```
Comment (hoechenberger) [inline mne_bids/path.py:1678]:
```suggestion
    ignore_descriptions = _ensure_tuple(ignore_descriptions)
```

**Comment by hoechenberger** [comment]
Failing tests all seem to be due to new set_montage() behavior upstream.

File: mne_bids/path.py [hunk 7]
```diff
@@ -288,36 +293,38 @@ class BIDSPath(object):
...
+            'subject': self.subject,
+            'session': self.session,
+            'task': self.task,
+            'acquisition': self.acquisition,
+            'run': self.run,
+            'processing': self.processing,
+            'space': self.space,
+            'recording': self.recording,
+            'split': self.split,
+            'description': self.description,
```
Comment (adam2392) [inline mne_bids/path.py:326]: Just a loose thought...

Is it worth differentiating "raw" entities and "derivative" entities? E.g. via a parameter passed to `entities`?

**Comment by alexrockhill** [comment]
> @alexrockhill can you have a look the doc build issue? It’s related to seeg example pray

Looks like it's passing now, happy to review though.

**Review by alexrockhill — APPROVED** [review]

**Review by agramfort — APPROVED** [review]

**Comment by agramfort** [comment]
@sappelhoff merge if happy 

🙏 @hoechenberger

File: mne_bids/path.py (hunk 7 above)
Comment (hoechenberger) [reply mne_bids/path.py:326]: I was thinking about this too at least for writing, but I'd say we can always add this later on... for now we had a specific use case where @agramfort couldnt read some OpenNeuro data with MNE-BIDS, and this PR fixes this :)

**Review by sappelhoff — APPROVED** [review]
Thanks Richard!

**Pull request merged**